### DISCUSSION TITLE
Only the first tab does not draw full width in fullscreen

### DISCUSSION THREAD

**Discussion by Strykar** [kovidgoyal/kitty#4877] [Q&A] [answered]
kitty is started via XDG autostart as:
```
[Desktop Entry]
Name=Weechat
GenericName=WeechatStartup
Comment=Start weechat in kitty term on login
Exec=/usr/bin/kitty --start-as fullscreen --instance-group=1 --config=/home/strykar/.config/kitty/weechatkitty.conf \
--session=/home/strykar/.config/kitty/weechat.session --listen-on=unix:/tmp/mykitty --directory=/home/strykar
StartupNotify=false
Terminal=false
Type=Application
X-GNOME-Autostart-enabled=true
```

Weechat, the IRC client is run in the first tab, with 4 more tabs opening with their apps.
Weechat always is sized to what the kitty terminal window _would be_ if started without the `--start-as fullscreen` option.
`Ctrl +  L` the keyboard shortcut for redrawing the whole window fixes the issue and now weechat runs full width.

It was suggested in the #weechat channel that this may be due to kitty not sending a sigwinch.
It was also suggested that weechat was being run before kitty started, which I'm unsure what to make of.
The 4th tab runs an htop like program, which works fine full width.
I could exec a delayed window redraw in weechat 20s after weechat is launched, but that is hackish.

Is this a kitty issue?
How can I fix this?

**Comment by page-down** [comment]
I also encountered similar problems.

> macOS: Allow opening URLs with kitty
> https://github.com/kovidgoyal/kitty/pull/4618#issuecomment-1031180550
>
> If the second file opened in the window launch --type=window (or invalid), then the timing may have caused the program in the first kitty window to fail to receive the window resize event to redraw.

The solution to the above issue is that if kitty already knows all the windows at startup (from open URLs, or session config), it can layout them first, and then run the programs all together.

For full screen:
Maybe kitty could be configured to run the program after full screen.
The issue can be reproduced on macOS and Linux with the following command.

```shell
kitty -o remember_window_size=no -o initial_window_width=80c --start-as fullscreen --hold sh -c 'c(){ tput cols; } ;c;sleep 0.1;c;sleep 0.5;c;'
```

> How can I fix this?

For now, my approach is to write wrapper scripts with delays.

**Comment by kovidgoyal** [comment, marked as answer]
This is simply a bug in the program. It is ignoring the SIGWINCH,
possibly because it is received during the program startup before the
signal handler is put in place. The program needs to put the sigwinch
handler in place *before* it checks the screen size for the first time.

kitty *always* sends sigwinch when resizing the window.

**Comment by page-down** [comment]
> This is simply a bug in the program. ...
> ... The program needs to put the sigwinch handler in place 
> *before* it checks the screen size for the first time.

https://github.com/weechat/weechat/blob/v3.4.1/src/gui/curses/normal/main.c#L42-L43
```c
    weechat_init (argc, argv, &gui_main_init);
    gui_main_loop ();
```

https://github.com/weechat/weechat/blob/v3.4.1/src/gui/curses/gui-curses-main.c
```c
void gui_main_init () {
    // ...
    gui_window_read_terminal_size ();
    // ...
}
void gui_main_loop () {
    // ...
    /* catch SIGWINCH signal: redraw screen */
    if (!weechat_headless)
        signal_catch (SIGWINCH, &gui_main_signal_sigwinch);
    // ...
}
```

So it appears to be related to these lines.

**Comment by Strykar** [comment]
@kovidgoyal Thank you, the bug was indeed in Weechat, fixed with https://weechat.org/files/temp/gui_signal_sigwinch_startup.patch
This thing has probably existed for the last 12 years, and curiously, the last time the sigwinch checks were moved around was to fix an issue around sigwinch :smile: 

@page-down Thank you for the pointers, it helped me get the attention of the Weechat devs sooner.
